### ISSUE TITLE
style guide: VLAs, function comments, true/false conditions, ...

### ISSUE THREAD

**Issue by justinmk** [neovim/neovim.github.io#53]
I believe we allow VLAs now, so we should modify the style guide to that effect.

Also since we now generate headers, we need to update the style guide passage regarding placement of function docstrings.

**Comment by justinmk** [comment]
More things to note in the style guide:

- `%zu` format specifier [#1729](https://github.com/neovim/neovim/pull/1729#discussion_r22436056)
- `PRId64` instead of `%ld`, etc. [#599](https://github.com/neovim/neovim/pull/599)

**Comment by fwalch** [comment]
* We autogenerate the header files, so the distinction between function declaration/definition comments does not apply https://github.com/neovim/neovim/pull/2284#discussion_r27564286

**Comment by justinmk** [comment]
- `strcmp` family of functions: prefer explicit comparison (`strcmp(...) == 0` vs `strcmp(...)` ?

**Comment by justinmk** [comment]
- prefer `(foo)` and `(!foo)` instead of `(foo != NULL)` and `(foo == NULL)`, where `foo` is a pointer.

**Comment by Pyrohh** [comment]
> prefer (foo) and (!foo) instead of (foo != NULL) and (foo == NULL), where foo is a pointer.

On one hand I think your suggestion is more idiomatic, but IMO it can make it harder to read code. If you encounter `(foo == NULL)` it's immediately obvious that it's a pointer, but `(foo)` is not. Of course `(foo)` could be in the presence of functions which imply it's a pointer, but this isn't necessarily obvious.

> prefer (foo == NUL) and (foo != NUL) where it is semantically appropriate.

Is this not already done?

**Comment by justinmk** [comment]
> On one hand I think your suggestion is more idiomatic, but IMO it can make it harder to read code. 

I'm neutral, but it seems that our most prolific contributors are in favor of it.

> Is this not already done?

You mean it's already in the style guide? I didn't check. We just need to update the style guide so people aren't confused.

**Comment by ZyX-I** [comment]
I think that `== NULL`/`!= NULL` checks are more readable as well. Also I cannot find where it is said that NULL *must* be zero:

> 7.17  Common definitions &lt;stddef.h&gt;
>
> The macros are `NULL` which expands to an implementation-defined null pointer constant; and …

Note **implementation-defined**.

Later it is explicitly said that null pointer may be *not* the pointer whose value is all-zeroes:

> 7.20.3.1 The `calloc` function
>
> The calloc function allocates space for an array of nmemb objects, each of whose size is size. The space is initialized to all bits zero.<sup>261)</sup>
>
> 261) Note that this need not be the same as the representation of floating-point zero or a null pointer constant.

. On the other side

> 6.3.2.3 Pointers
>
> An integer constant expression with the value 0, or such an expression cast to type `void *`, is called a *null pointer constant*. If a null pointer constant is converted to a pointer type, the resulting pointer, called a *null pointer*, is guaranteed to compare unequal to a pointer to any object or function.

I failed to find references which say that

1. Null pointer compares equal to zero. (The only condition on which converting to `_Bool` emits `false`, also condition on which various boolean expressions are assumed false.)
2. Null pointer converted to an integer type must emit zero (the above paragraph only says about the opposite: integer zero converted to a `void *` must be a null pointer).
3. Converting an integer to a pointer and back must emit the same result (it is only true for pointers in `ptr -> u?intptr_t -> ptr`, but not for integers in `u?intptr_t -> ptr -> u?intptr_t`: theoretically you may map one pointer to more then one different `u?intptr_t` integer value and use them randomly on conversion and this is going to work as long as no two pointers are converted to the same integer).
4. Converting a null pointer to `intptr_t`/`uintptr_t` must result in zero (even must be possible: the standard says that these types are optional).

So basically you *must not* use `if (ptr)` because *it is not portable*. If you know additional references which prove me wrong please show them.

**Comment by justinmk** [comment]
I think we're all aware of that, but any C compiler which defines NULL to anything other than 0 is guaranteed to not run any useful software. I don't feel strongly about it though.

**Comment by elmart** [comment]
@ZyX-I The standard may not guarantee that NULL representation is all zeros (though I don't know about any implementation that doesn't), but it is guaranteed that a null pointer (whatever its representation) will convert to false when evaluated in a boolean context. So, abbreviated form is perfectly safe.

**Comment by oni-link** [comment]
A faq for null pointers: http://c-faq.com/null/index.html. The problem with `if (ptr)` is answered in question 3.

**Comment by justinmk** [comment]
@oni-link  From 5.3:

> if(p)	is equivalent to	if(p != 0)

> and this is a comparison context, so the compiler can tell that the (implicit) 0 is actually a null pointer constant, and use the correct null pointer value. ... **The internal representation of a null pointer does *not* matter.**

The C-faq is often 95% complete, and then stops just short of spelling out a crucial detail. Is it saying if `NULL == 1`, the compiler will resolve the expression as `if(p != 1)`?

**Comment by oni-link** [comment]
> The C-faq is often 95% complete, and then stops just short of spelling out a crucial detail. Is it saying if NULL == 1, the compiler will resolve the expression as if(p != 1)?

In the explanation `NULL` is not used. Also `1` would not be an integral constant expression with the value 0 (so cannot be a null pointer), see @ZyX-I reference:
> 6.3.2.3 Pointers
An integer constant expression with the value 0, or such an expression cast to type void *, is called a null pointer constant. If a null pointer constant is converted to a pointer type, the resulting pointer, called a null pointer, is guaranteed to compare unequal to a pointer to any object or function

**Comment by elmart** [comment]
@justinmk No, it's saying the opposite. That the integer 0 will be converted to null pointer (whatever its representation) in order to be compared to p.

**Comment by ZyX-I** [comment]
On April 24, 2015 1:35:45 AM EAT, oni-link <notifications@github.com> wrote:
>> The C-faq is often 95% complete, and then stops just short of
>spelling out a crucial detail. Is it saying if NULL == 1, the compiler
>will resolve the expression as if(p != 1)?
>
>In the explanation `NULL` is not used. Also `1` would not be an
>integral constant expression with the value 0 (so cannot be a null
>pointer), see @ZyX-I reference:

It can. This section says that ... is called a null pointer constant ... converted to a pointer type ... called a null pointer. It does not say that this is *the only* way to get a null pointer. I do not see anywhere definition which says that two equal pointers can only be obtained from equal integers should they be obtained from integers.

Note that on a 32-bit system ((void *) ((uint64_t) 0xFFFFFFFF00000000LL)) will be a null pointer so it would be very stupid from the standard authors to require anything like this. But I also do not see why 0x1 *and* 0x0 converted to a pointer type cannot both be a null pointers. (I also do not know any systems where this actually happens though.)

**Comment by elmart** [comment]
Guys, I think this is degenerating into a kind of byzantine discussion.
From all the above, I think we can consider that using `if (p)` is perfectly safe.

**Comment by ZyX-I** [comment]
@elmart It is a discussion primary about coding style. We all know that “any C compiler which defines NULL to anything other than 0 is guaranteed to not run any useful software”, but if I have an additional point in defence to my position (pointer checks should be explicit) I will propagate it because primary point (“it is less readable”, expanded in one of the above comments (not mine)) is rather subjective.

**Comment by justinmk** [comment]
For `@returns` in doxygen comments, note `[out]` and `[allocated]` conventions.

https://github.com/neovim/neovim/pull/2470#discussion_r29104518

**Comment by justinmk** [comment]
Also need a caveat in the comment style recommendation regarding macro comments: https://github.com/neovim/neovim/pull/2578#discussion_r29609134

**Comment by dundargoc** [comment]
@justinmk Huh, I guess this is still relevant since the VLA section hasn't been updated. I'll create a PR now as to not forget it.

Is there anything else of use here to us? Skimmed through it quickly but mostly felt like I got a lecture on null pointers.

**Comment by justinmk** [comment]
@dundargoc i think this would be worth adding to `:help dev-style`: 

> For `@param` and `@returns` in doxygen comments, note the `[in]`, `[out]`, `[allocated]` conventions.

**Comment by dundargoc** [comment]
Plan to add it in [this PR](https://github.com/neovim/neovim/pull/15829).

**Comment by justinmk** [comment]
Style is now fully defined in the neovim/neovim repo in `src/uncrustify.cfg` and `:help dev-style`
